README.md
# dc-next
DC-SDK 2.x，DC-SDK 是基于 Cesium 进行二次开发的2、3D一体 WebGis 应用框架,该框架优化了 Cesium 的使用方式和增添了一些额外功能，旨在为开发者快速构建 WebGis 应用。🌎


CHANGES.md
# Change Log

### 1.16.0 - 2021-03-06

> 1. 升级 Cesium 到 1.79.1
> 2. 添加加载蒙层组件
> 3. 添加场景导出功能

### 1.15.2 - 2021-02-27

> 1. 开放 Cesium 部分常量
> 2. 抛出ready函数异常错误

### 1.15.1 - 2021-02-21

> 1. 添加相机控制的属性设置
> 2. 添加自定义天空盒
> 3. 完善天空盒（SkyBox) 的设置

### 1.15.0 - 2021-02-06

> 1. 升级 Cesium 到 1.78.0
> 2. 完善百度地图加载
> 2. 完善 ZoomController

### 1.14.3 - 2021-01-31

> 1. 添加百度地图对WGS84的支持
> 2. 添加高德地图对WGS84的支持
> 3. 添加百度、高德、WGS84坐标转换功能

### 1.14.2 - 2021-01-23

> 1. 添加覆盖物点位更新事件
> 2. 修改场景组件不销毁的问题

### 1.14.1 - 2021-01-16

> 1. 移入 CzmlLayer 和 KmlLayer
> 2. 完善场景组件的状态管理

### 1.14.0 - 2021-01-09

> 1. 升级 Cesium 到 1.77.0
> 2. 去除抛物线的点位认证
> 3. 完善场景组件禁用功能
> 4. 修改面积函数的无法使用的问题

### 1.13.3 - 2020-12-26

> 1. 添加鼠标事件中的WGS84坐标参数
> 2. 添加场景缩放控制组件
> 3. 完善坐标转换功能
> 4. 完善覆盖物右击菜单功能

### 1.13.2 - 2020-12-19

> 1. 完善图层管理功能
> 2. 完善 LocationBar 布局和事件相应功能
> 3. 完善 Position 以便支持各类类型的位置信息
> 4. 添加覆盖物对鼠标移动事件的支持

### 1.13.1 - 2020-12-12

> 1. 添加比例尺组件
> 2. 完善场景组件的初始化功能 
> 3. 开放部分场景属性
> 4. 添加覆盖物右击菜单，方便后期编辑

### 1.13.0 - 2020-12-05

> 1. 升级 Cesium 到 1.76
> 2. 移除场景效果相关函数
> 3. 完善地形移除函数
> 4. 完善右击菜单功能

### 1.12.3 - 2020-11-28

> 1. 移除 Cesium 默认地图 token 认证
> 2. 完善三维场景组件功能
> 3. 完善 parabola（抛物线)函数
> 4. 开放地图瓦片的设置

### 1.12.2 - 2020-11-23

> 1. 完善webpack打包
> 2. 完善圆材质旋转动画
> 3. 完善模型旋转动画
> 4. 开放场景部分设置功能

### 1.12.1 - 2020-11-14

> 1. 完善代码对低版本浏览器的支持
> 2. 添加创建视频节点函数
> 3. 添加部分Cesium内部函数

### 1.12.0 - 2020-11-07

> 1. 升级Cesium 到 1.75
> 2. 添加鼠标模式常量
> 3. 添加允许覆盖物穿透选择

### 1.11.2 - 2020-10-31

> 1. 重构天地图瓦片加载代码
> 2. 完善覆盖物添加和删除功能
> 3. 完善覆盖物圆的功能
> 4. 添加3dtiles的自定义shader

### 1.11.1 - 2020-10-20

> 1. 添加Tileset贴地函数
> 2. 完善DivIcon
> 3. 添加图元图层

### 1.11.0 - 2020-10-06

> 1. 升级Cesium 到 1.74
> 2. 添加根据类型创建地形的函数
> 3. 完善面要素中心点位

### 1.10.4 - 2020-09-26

> 1. 完善图层添加功能
> 2. 完善全局函数ready
> 3. 添加部分Cesium默认常量

### 1.10.2 - 2020-09-19

> 1. 完善罗盘功能
> 2. 修改框架包node下导入的方式
> 3. 修改DivIcon显示和隐藏功能
> 4. 完善覆盖物的属性设置

### 1.10.1 - 2020-09-12

> 1. 添加位置转换函数
> 2. 开放部分 Cesium.Viewer 属性
> 3. 完善气泡窗口和DivIcon的内容设置功能
> 4. 优化鹰眼功能

### 1.10.0 - 2020-09-05

> 1. 升级 Cesium 到 1.73
> 2. 添加百度路况地图
> 3. 修复鼠标滚轮事件无法监听的问题
> 4. 完善Dom工具类

### 1.9.4 - 2020-08-030

> 1. 添加标绘工具类
> 2. 添加图层组
> 3. 完善场景飞入目标函数

### 1.9.3 - 2020-08-22

> 1. 添加贝塞尔曲线、两点中心点的函数
> 2. 添加覆盖物支持Entity转换函数
> 3. 添加wms地图服务的支持

### 1.9.2 - 2020-08-15

> 1. 完善控制台输出说明
> 2. 修改说明文档

### 1.9.1 - 2020-08-11

> 1. 添加场景更新的事件定义
> 2. 添加百度影像图的支持
> 3. 完善webpack的配置

### 1.9.0 - 2020-08-04

> 1. 升级 Cesium 到 1.72
> 2. 完善代码注释功能

### 1.8.4 - 2020-07-31

> 1. 修改地图对比控件样式
> 2. 修复GeoJson图层不能转换为矢量图层的问题
> 3. 修复DomUtil中部分函数无法使用的问题
> 4. 修改Position中copy的调用方式

### 1.8.3 - 2020-07-27

> 1. 完善控件事件的初始化功能
> 2. 添加位置信息控件
> 3. 对 dc.core.min.js 进行混淆处理

### 1.8.2 - 2020-07-23

> 1. 修改高德矢量地图样式
> 2. 添加腾讯地图的样式参数
> 3. 修改相机翻转范围的问题

### 1.8.1 - 2020-07-20

> 1. 添加地球基础颜色的设置
> 2. 添加 3D 场景销毁功能

### 1.8.0 - 2020-07-02

> 1. 升级 Cesium 到 1.71
> 2. 完善鼠标事件
> 3. 默认组件的位置设置

### 1.7.6 - 2020-06-25

> 1. 修改全局函数 mixin 的限制
> 2. 添加设置标签函数到基础类

### 1.7.5 - 2020-06-17

> 1. 修改全局函数 use 无法使用的问题

### 1.7.4 - 2020-06-11

> 1. 升级 Cesium 到 1.70.1
> 2. 修改 webpack 配置

### 1.7.3 - 2020-06-09

> 1. 修改场景属性设置
> 2. 修改广告牌的尺寸设置问题

### 1.7.2 - 2020-06-06

> 1. 修改覆盖物属性设置和样式设置
> 2. 添加字符串转换为 Dom 节点函数
> 3. 支持多种点位参数格式
> 4. 添加 Icon 模块，将图片或者 svg 字符化管理
> 5. 添加 GeoJson 图层换为模型图层

### 1.7.1 - 2020-06-03

> 1. 删除 Cesium 无效文件
> 2. 修改覆盖物属性设置和样式设置

### 1.7.0 - 2020-06-03

> 1. 升级 Cesium 到 1.70
> 2. 修改场景参数设置，添加地表透明设置
> 3. 添加鼠标模式修改方法
> 4. 完善代码规范

### 1.6.3 - 2020-05-28

> 1. 修改函数 distance 的无法使用的问题
> 2. 添加抛物线函数
> 3. 添加 position 格式转换为 Turf 坐标格式函数

### 1.6.2 - 2020-05-25

> 1. 修改无法获取图层的问题

### 1.6.1 - 2020-05-19

> 1. 添加鼠标事件对覆盖物的验证
> 2. 修改 ViewerOption 调用错误
> 3. 修改覆盖物圆无法实例化的错误

### 1.6.0 - 2020-05-18

> 1. 修改无法添加和删除效果

### 1.5.2 - 2020-05-15

> 1. 修改鼠标事件的类型
> 2. 添加覆盖物的鼠标移除和移入的事件

### 1.5.1 - 2020-05-12

> 1. 重新组织目录结构，添加 ES6 的模块支持
> 2. 去除 Overlay、Plugins、Plot 的模块，将模块拆分成单独包
> 3. 添加点位和场景的验证函数
> 4. 框架中添加 mixin 和 use 全局函数

### 1.4.2 - 2020-05-08

> 1. 添加许可说明
> 2. 添加 npm 发布支持
> 3. 修改各类效果的添加方式

### 1.4.0 - 2020-05-06

> 1. 添加地图 subdomain 的限制
> 2. 修改图层、覆盖物添加到父类容器的函数
> 3. 添加屏幕坐标转 WGS84 函数
> 4. 升级 Cesium 框架到 1.69
> 5. 修复覆盖物获取问题

### 1.3.0 - 2020-04-27

> 1. DC.Math 添加计算长度、面积、中心点等函数
> 2. 场景中添加 ZoomPosition 方法
> 3. 添加图层和要素状态变化的钩子函数

### 1.2.0 - 2020-04-24

> 1. 重构热区图层
> 2. 添加覆盖物 id 属性，可通过 id 获取覆盖物
> 3. 规范方法名的编写格式

### 1.1.0 - 2020-04-22

> 1. 将 Cesium 单独拆分，作为 dc-base
> 2. 修复全局鼠标无法是使用的 bug
> 3. 修改除了 dc-base 外模块包的导入方式
> 4. 修改获取原生 Cesium 的方式

### 1.0.0 - 2020-04-18

> 1. 完成 3D 场景的创建
> 2. 3D 场景的设置，包括：相机的设置，地球的设置，场景的设置等
> 3. 图层的管理，包括：状态、添加、删除
> 4. 覆盖物的管理，包括：状态、事件、添加、删除
> 5. 效果的管理
> 6. 基本后期的添加


webpack.conf.js
/*
 * @Author: Caven
 * @Date: 2020-01-18 18:22:23
 * @Last Modified by: Caven
 * @Last Modified time: 2020-06-04 09:42:54
 */

const path = require('path')
const webpack = require('webpack')
const MiniCssExtractPlugin = require('mini-css-extract-plugin')
const OptimizeCssAssetsPlugin = require('optimize-css-assets-webpack-plugin')
const CopywebpackPlugin = require('copy-webpack-plugin')
const cesiumBuild = 'node_modules/cesium/Build/Cesium'

let cesiumCopyPlugin = [
  new CopywebpackPlugin([
    { from: path.join(cesiumBuild, 'Assets'), to: 'resources/Assets' }
  ]),
  new CopywebpackPlugin([
    { from: path.join(cesiumBuild, 'Workers'), to: 'resources/Workers' }
  ]),
  new CopywebpackPlugin([
    { from: path.join(cesiumBuild, 'ThirdParty'), to: 'resources/ThirdParty' }
  ])
]

function resolve(dir) {
  return path.join(__dirname, '.', dir)
}

module.exports = env => {
  const IS_PROD = (env && env.production) || false
  const publicPath = IS_PROD ? '/' : '/'
  let plugins = [
    ...cesiumCopyPlugin,
    new MiniCssExtractPlugin({
      filename: IS_PROD ? '[name].min.css' : '[name].css',
      allChunks: true
    }),
    new webpack.DefinePlugin({
      CESIUM_BASE_URL: JSON.stringify('./libs/dc-sdk/resources/')
    })
  ]
  if (IS_PROD) {
    plugins.push(new OptimizeCssAssetsPlugin())
    plugins.push(new webpack.NoEmitOnErrorsPlugin())
  }
  return {
    entry: {
      'dc.base': ['base'],
      'dc.core': ['entry', 'theme']
    },
    devtool: IS_PROD ? false : 'cheap-module-eval-source-map',
    output: {
      filename: IS_PROD ? '[name].min.js' : '[name].js',
      path: path.resolve(__dirname, 'dist'),
      publicPath: publicPath,
      sourcePrefix: ''
    },
    amd: {
      toUrlUndefinded: true
    },
    node: {
      fs: 'empty'
    },
    module: {
      unknownContextCritical: false,
      rules: [
        {
          test: /\.js$/,
          exclude: /node_modules/,
          loader: 'babel-loader',
          options: {
            presets: ['@babel/preset-env'],
            plugins: ['@babel/transform-runtime'],
            compact: false,
            ignore: ['checkTree']
          }
        },
        {
          test: /\.css$/,
          use: [
            MiniCssExtractPlugin.loader,
            {
              loader: 'css-loader'
            },
            {
              loader: 'sass-loader'
            }
          ]
        },
        {
          test: /\.scss$/,
          use: [
            MiniCssExtractPlugin.loader,
            {
              loader: 'css-loader'
            },
            {
              loader: 'sass-loader'
            }
          ]
        },
        {
          test: /\.(png|jpg|gif|svg|eot|ttf|woff|woff2)$/,
          loader: 'url-loader',
          options: {
            limit: 20000
          }
        },
        {
          test: /\.glsl$/,
          loader: 'webpack-glsl-loader'
        }
      ]
    },
    resolve: {
      extensions: ['.js', '.json', '.css'],
      alias: {
        '@': resolve('src'),
        base: './src/base/index.js',
        entry: './src/core/index.js',
        theme: './src/themes/index.js',
        cesium: path.resolve(__dirname, cesiumBuild)
      }
    },
    plugins
  }
}
